popx/transaction.go

```diff
@@ -35,7 +36,13 @@ func Transaction(ctx context.Context, connection *pop.Connection, callback func(
 				return err
 			}
 
+			attempt := 0
 			return crdb.ExecuteInTx(ctx, sqlxTxAdapter{transaction.TX.Tx}, func() error {
+				attempt++
+				if attempt > 1 {
+					caller := caller()
+					TransactionRetries.WithLabelValues(caller).Inc()
```

Comment:
Does this mean that the counters are per caller and not per calls? So two retries for one tx, or one retry each for two txs, all from the same caller, result in a retry counter of two? Is that desired?

Reply:
Yes, the counter is per caller. Otherwise you'd get a cardinality explosion in the labels (very common problem in Prometheus-style metrics). The idea is to have a dashboard graphing the number of transaction retries, grouped by the calling function (think `persister.CreateIdentity`, `session.Create` etc). That allows you to identify and monitor problematic code paths and fix them one-by-one.

I don't think having per-call events for retries is particularly helpful. They're a pretty normal occurence every once in a while, and aggregating them by caller gives us all the info we need without having to log. Logging would require injecting a logger, etc etc.